src/collection/tx_data/historical_data_collector.py

```diff
@@ -0,0 +1,78 @@
+"""
+Collector of raw transaction data for historical transactions.
+Dependent on `transactions` db table as its source of blocks to fetch. All PPKs should be known.
+
+
+Fetch all transactions from this block, filter by PPKs of all protocols (i.e., Solend, Mango, etc).
+
+1. Obtains several (`BATCH_SIZE`) oldest slot numbers where there are relevant transactions
+ that do not have transaction data stored in database.
+2. Concurrently fetch blocks from these slots.
+3. Filter all transactions in fetched blocks by PPKs of all lending protocols to select relevant transactions.
+4. For each relevant transaction:
+    - create new record in transactions table if not yet exists.
+    - assign transaction data to existing records.
+5. Repeat 1-4 indefinitely.
+
+If being restarted - nothing changes.
+"""
+import asyncio
+import logging
+import os
+
+import db
+from src.collection.tx_data.collector import TXFromBlockCollector
+
+LOG = logging.getLogger(__name__)
+BATCH_SIZE = 100
+OFFSET = os.getenv("OFFSET", "0")
+
+
+class HistoricalTXCollector(TXFromBlockCollector):
+
+    @property
+    def COLLECTION_STREAM(self) -> db.CollectionStreamTypes:  # pylint: disable=invalid-name
+        """ Property returning stream type."""
+        return db.CollectionStreamTypes.HISTORICAL
+
+    def _get_assigned_blocks(self) -> None:
+        """
+        Get `BATCH_SIZE` of unique block numbers we need to fetch and get transaction data from.
+        :return:
+        """
+        # TODO: make it more flexible. we want to be able to get assignments indirectly  # pylint: disable=W0511
+        #  from db so several collectors can serve at once / or change status of tx while fetching tx_raw
+        # Fetch first n blocks that contain transactions without tx_raw.
+        with db.get_db_session() as session:
+            distinct_slots = session.query(
+                db.TransactionStatusWithSignature.slot
+            ).filter(
+                db.TransactionStatusWithSignature.transaction_data.is_(None)
+            ).distinct().offset(int(OFFSET)).limit(BATCH_SIZE*3).subquery()
```

Comment:
What does the offset do?

Reply:
We have several containers doing collection. Each takes the oldest `$BATCH_SIZE` (now 100) blocks from database. To ensure that containers are not collecting the same block numbers, each of containers has some OFFSET, i.e does not take last 100 blocks but from `OFFSET` to `OFFSET+100`.

i add comment next to the definition of variable:
```
# Defines offset of `HistoricalTXCollector._get_assigned_blocks method.
# In case when several historical data collector are run simultaneously, different offsets will ensure
# that these collectors don't get same block numbers in assignment.
OFFSET = os.getenv("OFFSET", "0")
```

Reply:
So when you run several containers at once, you set different `OFFSET` for each of them, right? 

If so, we could choose the offset randomly (e.g. random integer from <1, 100>, multiplied by `BATCH_SIZE`), so that we don't have to do that manually.

Reply:
yes, usually it is 500, 1000 etc
we want to keep each containers' selection range as far from each other as possible. 

Its a bit off workaround, because we need to speed up collection. this kind of intervention is manual in general (you have to run new container)